tests/storage/test_user_directory.py

```diff
@@ -584,6 +584,36 @@ def test_search_user_dir_dotted_dotless_i_case_insensitivity(self) -> None:
 
     test_search_user_dir_dotted_dotless_i_case_insensitivity.skip = "not supported"  # type: ignore
 
+    @override_config({"user_directory": {"search_all_users": True}})
+    def test_search_user_dir_unicode_normalization(self) -> None:
+        """Tests that a user can look up another user by searching for their name with
+        either composed or decomposed accents.
+        """
+        AMELIE = "@someuser:example.org"
+
+        expected_matches = [
+            # (search_term, display_name)
+            ("Ame\u0301lie", "Amélie"),
+            ("Amélie", "Ame\u0301lie"),
+        ]
+
+        for search_term, display_name in expected_matches:
+            self.get_success(
+                self.store.update_profile_in_user_dir(AMELIE, display_name, None)
+            )
+
+            r = self.get_success(self.store.search_user_dir(ALICE, search_term, 10))
+            self.assertFalse(r["limited"])
+            self.assertEqual(
+                1,
+                len(r["results"]),
+                f"searching for {search_term!r} did not match {display_name!r}",
```

Comment:
Just curious what the bang operator here is doing - I've only ever seen it as part of a not-equal operator, and I can't quite parse out what "!r" is doing here in the error message.

Reply:
This is part of the `str.format` language, see https://docs.python.org/3/library/string.html#formatstrings or  https://pyformat.info/#conversion_flags

It means we're printing out `repr(x)` rather than `str(x)`.

Reply:
Ahh that makes sense, I got confused because the result of the `search_user_dir` call right above is also `r` so I was trying to figure out how played a role but turns out it was just irrelevant information...

Reply:
Thanks!